### PR TITLE
Add ensembles

### PR DESCRIPTION
**Pull request by cisprague**
We now have a working prototype of ensembles in a tutorial notebook. Further development will result in complexity being transitioned from the tutorial notebook to the codebase.

Features:
- [x] Ensemble of emulators (e.g. MLPs or other ensembles).
- [x] Dropout stochastic emulator.

Other things:
- When calling `ensemble.fit`, it would be nice to see some feedback on the progress of `emulator.fit` for each of the emulators in the ensemble. Defining a custom `__repr__` or casting all emulators as dataclasses are options for this. Dataclasses are helpful for this (especially the recursion needed to traverse all sub emulators).
- Need to think if iter-sample covariance is needed, like in GPs.
- The MLP defined in the tutorial notebook should be replaced by the one resulting from #559.

### CODE REVIEW

**Comment by sgreenbury** [comment]
>     * When calling `ensemble.fit`, it would be nice to see some feedback on the progress of `emulator.fit` for each of the emulators in the ensemble. Defining a custom `__repr__` or casting all emulators as dataclasses are options for this. Dataclasses are helpful for this (especially the recursion needed to traverse all sub emulators).

This sounds great, though I am not sure if we can make emulators dataclasses since the way we are subclassing e.g. `nn.Module`, `ExactGP` and others?

>     * Need to think if iter-sample covariance is needed, like in GPs.

It would be good to discuss whether to have this feature! If it is, updating to [`MultitaskMultivariateNormal`](https://docs.gpytorch.ai/en/v1.6.0/_modules/gpytorch/distributions/multitask_multivariate_normal.html) (`GaussianLike` in the types) should work for this. Also relates to the #471.

>     * The MLP defined in the tutorial notebook should be replaced by the one resulting from [Add MLP #559](https://github.com/alan-turing-institute/autoemulate/pull/559).

Sounds good! I've updated #559 to have dropout as an option (but only during training) - it would be good to discuss the stochastic emulator case at inference time.

File: tests/experimental/test_experimental_ensemble.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        emulators.append(emulator)
+
+    # Ensemble emulator
+    emulator = Ensemble(emulators)
+    emulator.fit(x_train, y_train)
+
+    # Test
+    x_test = torch.linspace(0.0, 15.0, steps=1000).reshape(-1, 1)
+    y_test_hat = emulator.predict(x_test)
+    assert isinstance(y_test_hat, torch.distributions.MultivariateNormal)
```
Comment (sgreenbury) [inline tests/experimental/test_experimental_ensemble.py:28]:
```suggestion
    assert isinstance(y_test_hat, GaussianLike)
```

File: tests/experimental/test_experimental_ensemble.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    # Dropout Ensemble
+    emulator = MLP(x_train, y_train, layer_dims=[100, 100], lr=1e-3, dropout_prob=0.10)
+    emulator.epochs = 10
+    emulator = DropoutEnsemble(emulator, n_samples=20, jitter=1e-7)
+    emulator.fit(x_train, y_train)
+
+    # Test
+    x_test = torch.linspace(0.0, 15.0, steps=1000).reshape(-1, 1)
+    y_test_hat = emulator.predict(x_test)
+    assert isinstance(y_test_hat, torch.distributions.MultivariateNormal)
```
Comment (sgreenbury) [inline tests/experimental/test_experimental_ensemble.py:50]:
```suggestion
    assert isinstance(y_test_hat, GaussianLike)
```

File: autoemulate/experimental/emulators/ensemble.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        device = x.device
+        means: list[Tensor] = []
+        covs: list[Tensor] = []
+
+        # Outputs from each emulator
+        for e in self.emulators:
+            out = e.predict(x)
+            if isinstance(out, GaussianLike):
+                mu_i = out.mean.to(device)  # (batch_size, n_dims)
+                sigma_i = out.covariance_matrix.to(
```
Comment (sgreenbury) [inline autoemulate/experimental/emulators/ensemble.py:83]: For the pyright issue:
```suggestion
                assert isinstance(out.covariance_matrix, TensorLike)
                sigma_i = out.covariance_matrix.to(
```
Not sure the reason for the difference locally and the in the CI.

File: autoemulate/experimental/emulators/ensemble.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    Ensemble emulator that aggregates multiple Emulator instances and returns
+    a MultivariateNormal representing the ensemble posterior.
+    Note that an Emulator instance may also be an Ensemble itself.
+
+    Parameters
+        Emulators: list[Emulator] List of fitted emulator instances.
+        Each member's `predict(x: Tensor)`
+        must return either:
+            - TensorLike (treated as mean)
+            - MultivariateNormal (with `.mean` and `.covariance_matrix`)
```
[the remark this answers is not in the document]
Comment (cisprague) [reply autoemulate/experimental/emulators/ensemble.py:31]: Yes, will do that. Also, in general (maybe this should be a separate issue), how do we want to document public (e.g. `.predict`) and private methods (e.g. `_predict`)?

File: docs/tutorials/08_ensembles.ipynb
Comment (sgreenbury) [inline docs/tutorials/08_ensembles.ipynb:1]: Update number to 09

**Review by sgreenbury — APPROVED** [review]
This looks really great @cisprague, thanks! Just a couple of small comments as discussed above and a new issue to be opened for further revisions to the notebook but otherwise looks good to merge.

File: docs/tutorials/08_ensembles.ipynb
Comment (sgreenbury) [reply docs/tutorials/08_ensembles.ipynb:1]: Resolved as #592 renumbers
